File: android/core/designsystem/src/main/kotlin/com/devora/mencare/core/designsystem/component/BrandInputs.kt
package com.devora.mencare.core.designsystem.component

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp
import com.devora.mencare.core.designsystem.R
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.ErrorRed
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.OliveWood
import com.devora.mencare.core.designsystem.theme.Stone
import com.devora.mencare.core.designsystem.theme.TextMuted

internal val FieldShape = RoundedCornerShape(14.dp)
internal val FieldHeight = 52.dp

/** Uppercase letter-spaced label sitting above an input, as in the design mockup. */
@Composable
fun FieldLabel(text: String, modifier: Modifier = Modifier) {
    Text(
        text.uppercase(),
        style = MaterialTheme.typography.labelMedium.copy(fontSize = 11.sp, letterSpacing = 0.14.em),
        color = Ink,
        modifier = modifier,
    )
}

private val FieldTextSize = 15.sp

/** Filled input on the Stone surface — the resting state of every text field in the app. */
@Composable
fun FilledTextField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    placeholder: String? = null,
    leadingIcon: ImageVector? = null,
    prefix: String? = null,
    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
    error: String? = null,
    enabled: Boolean = true,
    height: Dp = FieldHeight,
    outlined: Boolean = false,
    trailing: @Composable (() -> Unit)? = null,
) {
    Column(modifier.fillMaxWidth()) {
        if (label.isNotBlank()) {
            FieldLabel(label)
            Spacer(Modifier.height(6.dp))
        }
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            enabled = enabled,
            singleLine = true,
            keyboardOptions = keyboardOptions,
            textStyle = MaterialTheme.typography.bodyLarge.copy(color = Ink, fontSize = FieldTextSize),
            cursorBrush = SolidColor(OliveWood),
            modifier = Modifier.fillMaxWidth().height(height),
        ) { field ->
            Row(
                modifier = Modifier
                    .fillMaxSize()
                    .clip(FieldShape)
                    .background(if (outlined) Bone else Stone)
                    .then(
                        if (outlined) Modifier.border(1.5.dp, OliveWood, FieldShape) else Modifier,
                    )
                    .padding(horizontal = 15.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                if (leadingIcon != null) {
                    Icon(
                        leadingIcon,
                        contentDescription = null,
                        tint = Ink.copy(alpha = 0.45f),
                        modifier = Modifier.size(17.dp),
                    )
                    Spacer(Modifier.width(10.dp))
                }
                if (prefix != null) {
                    Text(
                        prefix,
                        style = MaterialTheme.typography.titleSmall,
                        color = Ink,
                    )
                    Spacer(Modifier.width(11.dp))
                    Box(
                        Modifier
                            .width(1.dp)
                            .height(20.dp)
                            .background(Ink.copy(alpha = 0.14f)),
                    )
                    Spacer(Modifier.width(11.dp))
                }
                Box(Modifier.weight(1f)) {
                    if (value.isEmpty() && placeholder != null) {
                        Text(
                            placeholder,
                            style = MaterialTheme.typography.bodyLarge.copy(fontSize = FieldTextSize),
                            color = Ink.copy(alpha = 0.4f),
                        )
                    }
                    field()
                }
                if (trailing != null) {
                    Spacer(Modifier.width(10.dp))
                    trailing()
                }
            }
        }
        FieldMessage(error)
    }
}

/** Password input: outlined in Olive Wood with an inline show/hide toggle. */
@Composable
fun BorderedPasswordField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    error: String? = null,
    height: Dp = FieldHeight,
) {
    var visible by rememberSaveable { mutableStateOf(false) }
    Column(modifier.fillMaxWidth()) {
        FieldLabel(label)
        Spacer(Modifier.height(6.dp))
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
            visualTransformation = if (visible) VisualTransformation.None else PasswordVisualTransformation(),
            textStyle = MaterialTheme.typography.bodyLarge.copy(
                color = Ink,
                fontSize = FieldTextSize,
                letterSpacing = if (visible) 0.em else 0.26.em,
            ),
            cursorBrush = SolidColor(OliveWood),
            modifier = Modifier.fillMaxWidth().height(height),
        ) { field ->
            Row(
                modifier = Modifier
                    .fillMaxSize()
                    .clip(FieldShape)
                    .background(Bone)
                    .border(1.5.dp, if (error != null) ErrorRed else OliveWood, FieldShape)
                    .padding(horizontal = 15.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Box(Modifier.weight(1f)) { field() }
                val toggleLabel = stringResource(
                    if (visible) R.string.ds_hide_password else R.string.ds_show_password,
                )
                Text(
                    stringResource(
                        if (visible) R.string.ds_hide_password_short else R.string.ds_show_password_short,
                    ),
                    style = MaterialTheme.typography.titleSmall,
                    color = OliveWood,
                    modifier = Modifier
                        .clickable(onClickLabel = toggleLabel) { visible = !visible }
                        .padding(start = 10.dp),
                )
            }
        }
        FieldMessage(error)
    }
}

/**
 * Riga sotto un campo: l'errore in rosso, altrimenti l'aiuto in grigio.
 * È l'unico posto in cui si scrive un messaggio sotto a un input.
 */
@Composable
fun FieldMessage(
    error: String?,
    modifier: Modifier = Modifier,
    helper: String? = null,
) {
    val text = error ?: helper ?: return
    Text(
        text,
        style = MaterialTheme.typography.bodySmall,
        color = if (error != null) ErrorRed else TextMuted,
        modifier = modifier.padding(top = 6.dp),
    )
}

File: ios/MenCare/Core/DesignSystem/Component/BrandInputs.swift
import SwiftUI

private let fieldCorner: CGFloat = 14
private let fieldHeight: CGFloat = 52
private let fieldTextSize: CGFloat = 15

/// Uppercase letter-spaced label sitting above an input, as in the design mockup.
struct FieldLabel: View {
    let text: String

    var body: some View {
        Text(text.uppercased())
            .font(Typo.jost(11, weight: .medium))
            .kerning(1.5)
            .foregroundStyle(Color.ink)
    }
}

/// Filled input on the Stone surface — the resting state of every text field in the app.
struct FilledTextField<Trailing: View>: View {
    let label: String
    @Binding var text: String
    var placeholder: String?
    var leadingSystemImage: String?
    var prefix: String?
    var keyboard: UIKeyboardType = .default
    var autocapitalization: TextInputAutocapitalization = .never
    var error: String?
    /// Riga grigia sotto il campo quando non c'è errore (es. il formato atteso).
    var helper: String?
    var enabled: Bool = true
    var height: CGFloat = fieldHeight
    var outlined: Bool = false
    /// Passata dai campi condivisi che devono normalizzare il valore al blur
    /// (il telefono): `.focused` funziona solo sulla TextField vera.
    var focus: FocusState<Bool>.Binding?
    @ViewBuilder var trailing: Trailing

    init(
        label: String,
        text: Binding<String>,
        placeholder: String? = nil,
        leadingSystemImage: String? = nil,
        prefix: String? = nil,
        keyboard: UIKeyboardType = .default,
        autocapitalization: TextInputAutocapitalization = .never,
        error: String? = nil,
        helper: String? = nil,
        enabled: Bool = true,
        height: CGFloat = fieldHeight,
        outlined: Bool = false,
        focus: FocusState<Bool>.Binding? = nil,
        @ViewBuilder trailing: () -> Trailing = { EmptyView() }
    ) {
        self.label = label
        self._text = text
        self.placeholder = placeholder
        self.leadingSystemImage = leadingSystemImage
        self.prefix = prefix
        self.keyboard = keyboard
        self.autocapitalization = autocapitalization
        self.error = error
        self.helper = helper
        self.enabled = enabled
        self.height = height
        self.outlined = outlined
        self.focus = focus
        self.trailing = trailing()
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            if !label.isEmpty {
                FieldLabel(text: label)
            }
            HStack(spacing: 10) {
                if let leadingSystemImage {
                    Image(systemName: leadingSystemImage)
                        .font(.system(size: 15))
                        .foregroundStyle(Color.ink.opacity(0.45))
                }
                if let prefix {
                    Text(prefix).font(Typo.titleSmall).foregroundStyle(Color.ink)
                    Rectangle()
                        .fill(Color.ink.opacity(0.14))
                        .frame(width: 1, height: 20)
                }
                textField
                    .font(Typo.jost(fieldTextSize))
                    .foregroundStyle(Color.ink)
                    .tint(Color.oliveWood)
                    .keyboardType(keyboard)
                    .textInputAutocapitalization(autocapitalization)
                    .autocorrectionDisabled()
                    .disabled(!enabled)
                trailing
            }
            .padding(.horizontal, 15)
            .frame(height: height)
            .background(RoundedRectangle(cornerRadius: fieldCorner).fill(outlined ? Color.bone : Color.stone))
            .overlay(
                RoundedRectangle(cornerRadius: fieldCorner)
                    .strokeBorder(
                        error != nil ? Color.errorRed : Color.oliveWood,
                        lineWidth: 1.5
                    )
                    .opacity(outlined || error != nil ? 1 : 0)
            )
            FieldMessageRow(error: error, helper: helper)
        }
    }

    @ViewBuilder
    private var textField: some View {
        if let focus {
            TextField(placeholder ?? "", text: $text).focused(focus)
        } else {
            TextField(placeholder ?? "", text: $text)
        }
    }
}

/// Password input: outlined in Olive Wood with an inline show/hide toggle.
struct BorderedPasswordField: View {
    let label: String
    @Binding var text: String
    var error: String?
    var helper: String?
    var height: CGFloat = fieldHeight

    @State private var visible = false

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            FieldLabel(text: label)
            HStack(spacing: 10) {
                Group {
                    if visible {
                        TextField("", text: $text)
                    } else {
                        SecureField("", text: $text)
                    }
                }
                .font(Typo.jost(fieldTextSize))
                .kerning(visible ? 0 : 3)
                .foregroundStyle(Color.ink)
                .tint(Color.oliveWood)
                .textInputAutocapitalization(.never)
                .autocorrectionDisabled()
                Button {
                    visible.toggle()
                } label: {
                    Text(visible ? String(localized: "ds_hide_password_short") : String(localized: "ds_show_password_short"))
                        .font(Typo.titleSmall)
                        .foregroundStyle(Color.oliveWood)
                }
                .buttonStyle(.plain)
            }
            .padding(.horizontal, 15)
            .frame(height: height)
            .background(RoundedRectangle(cornerRadius: fieldCorner).fill(Color.bone))
            .overlay(
                RoundedRectangle(cornerRadius: fieldCorner)
                    .strokeBorder(error != nil ? Color.errorRed : Color.oliveWood, lineWidth: 1.5)
            )
            FieldMessageRow(error: error, helper: helper)
        }
    }
}

/// Label above a tappable value: the date/time tiles used across the editors.
/// Takes the whole width it is offered, so tiles in a row come out equal.
struct PickerTile: View {
    let label: String
    let value: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            VStack(alignment: .leading, spacing: 3) {
                Text(label.uppercased())
                    .font(Typo.jost(9, weight: .medium))
                    .kerning(0.9)
                    .foregroundStyle(Color.textMuted)
                    .lineLimit(1)
                    .minimumScaleFactor(0.8)
                Text(value)
                    .font(Typo.jost(15, weight: .medium))
                    .foregroundStyle(Color.ink)
                    .lineLimit(1)
                    .minimumScaleFactor(0.8)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 14)
            .padding(.vertical, 11)
            .background(RoundedRectangle(cornerRadius: 16).fill(Color.stone))
        }
        .buttonStyle(.plain)
    }
}
